File: SBiswas/Problems/Milestone1/Problem30.cpp
#include<iostream>
#include<vector>
#include<unordered_map>

using namespace std;


//  Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 };

class Solution {
public:

    TreeNode* treeBuilder(vector<int>& nums, int left, int right)
    {
        if(left > right)
        {
            return nullptr;
        }

        int mid = (left + right)/2;

        TreeNode* root = new TreeNode(nums[mid]);
        root->right = treeBuilder(nums, mid+1, right);
        root->left = treeBuilder(nums, left, mid-1);

        return root;
    }

    TreeNode* sortedArrayToBST(vector<int>& nums) {
        int size = nums.size();

        int left = 0;
        int right = size-1;

        TreeNode* root = treeBuilder(nums, left, right);

        return root;

    }
};

File: Hritwik/Problems/Milestone1/Problem30.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def sortedArrayToBST(self, nums):
        if not nums:
            return None
        # Get the middle element of the sorted element to divide the array into two parts
        new_node=TreeNode(nums[len(nums)//2]) 
        new_node.left=self.sortedArrayToBST(nums[:len(nums)//2])
        new_node.right=self.sortedArrayToBST(nums[len(nums)//2+1:])
        return new_node

solution = Solution()
nums = [-10,-3,0,5,9]
response = solution.sortedArrayToBST(nums)

